### ISSUE TITLE
Support for .net Core and EF core?

### ISSUE THREAD

**Issue by Eneuman** [Azure/elastic-db-tools#118]
Hi!

Will there be support for .Net Core and Entity Framework 7?
If so, when? :)

**Comment by jaredmoo** [comment]
Hi, thanks for the request! We have no plans to implement this in the near future, but the code is open-source so if you'd like to give it a try yourself I'd be happy to work with you on getting it merged. :)

**Comment by andrelmp** [comment]
i tryed, but is pretty impossible now, .net core RC1 doesnt have `IDataReader` and `IDataRecord` interfaces also `SqlDataReader.GetSchemaTable()` is missing.

I will look into again when RC2 comes out

**Comment by abatishchev** [comment]
Any chance you could take a look at EF Core 1.0 please? Also I'd rename the title, from EF7 to EF Core.

**Comment by pristinesource** [comment]
https://github.com/pristinesource/ElasticDatabaseToolsNetCore

**Comment by jaredmoo** [comment]
Awesome, thanks @pristinesource ! Our build situation is a little weird but I'll see if I can slip the .xproj etc that you created into our codebase so we don't end up with 2 parallel codebases.

**Comment by pristinesource** [comment]
Awesome!  There were certainly a lot more changes than just the project file.  I will do 2 commits, one reverting all changes and another reapplying all changes to you can see everything I did in one commit.  The way Net Core projects work though, I had to create a whole new project and copy the source files to it instead of modifying the existing project.  I will add another comment here after I make those 2 commits with a link to the diff.

**Comment by pristinesource** [comment]
@jaredmoo, I have created a single commit that has all of the changes.  I realized that a huge part of it was the namespace refactor, but if you ignore that, you can see all of the work I did to get this library to work with .Net Core.  I hope this helps and I would LOVE to see an official .Net Core version soon :)

https://github.com/pristinesource/ElasticDatabaseToolsNetCore/commit/4d4917f6385ae1062d2a4db93a5fdfd035db2fda

**Comment by jaredmoo** [comment]
Hey, I'm working through the changes and it all makes sense except for the change to MultiShardDataReader.Close() to make it not override. Do you understand what's going on here? [DbDataReader has `virtual void Close()` defined](https://github.com/dotnet/corefx/blob/master/src/System.Data.Common/src/System/Data/Common/DbDataReader.cs#L288), I don't understand why it can't be overridden.

**Comment by pristinesource** [comment]
@jaredmoo: That is very strange, when I try to add "override" back to the method declaration in VS2015 it gives the error that there is no valid method to override.  When I look at the API docs for the DbDataReader it *does* say that there is a Close() method and it is virtual <https://docs.microsoft.com/en-us/dotnet/core/api/system.data.common.dbdatareader#System_Data_Common_DbDataReader_Close>, so I am not sure what the dependency is on my end.  Are you seeing any similar behavior when you open my project directly?
![screenshot 2016-09-21 16 26 55](https://cloud.githubusercontent.com/assets/197171/18729790/d1b788b8-8018-11e6-856e-9a8a7acf5231.png)
![screenshot 2016-09-21 16 29 21](https://cloud.githubusercontent.com/assets/197171/18729791/d1b969d0-8018-11e6-9125-c70c8e35cb86.png)

**Comment by jaredmoo** [comment]
Yes I see the same thing. What's also weird is that when I press F12 on DbDataReader, VS shows me the metadata with no `.Close` method. So I think that somehow DbDataReader class that is being referenced is not the same as the source that I just linked. Very baffling.

**Comment by jaredmoo** [comment]
Also these errors makes no sense. DbDataReader inherits IDataReader, which contains .Close().

![error](https://cloud.githubusercontent.com/assets/3248887/18730042/2118e452-8009-11e6-8a24-5bb4a8674d9b.png)

**Comment by pristinesource** [comment]
@jaredmoo:  It has confused me a lot also even with other stuff in .Net Core.  Take for example if you attempt to open the referenced DLL (the comment at the top of F12) in ILSpy it only gives the structure (which matches F12 output and that is correct because it is a *reference* DLL), but I cannot find the real library that matches the reference DLL.  I am not sure what the true referenced assembly is in this case, if it is the corefx library then it *does have* the Close() method (according to ILSpy), but that is not what VS2015 and the compiler see/use.

**Comment by jaredmoo** [comment]
Hmm yeah maybe the reference assembly that VS is using is not correct? I'll try contacting .net core folks. btw my working branch is https://github.com/jaredmoo/elastic-db-tools/tree/NetCore

**Comment by pristinesource** [comment]
@jaredmoo: Awesome! Then maybe all of the calls to *.Close() can be uncommented too ;)

**Comment by jaredmoo** [comment]
https://github.com/dotnet/corefx/issues/11949

**Comment by jaredmoo** [comment]
Apparently Close() should be replaced with Dispose(), see dotnet/corefx#11949. Since there are a lot of issues in MSQ, I totally disabled MSQ for now. The good news is that SMM now builds successfully in https://github.com/jaredmoo/elastic-db-tools/tree/NetCore . Next step is to get the SMM tests to build so that we can find out if these compile fixes actually produced an assembly that works good :)

**Comment by jaredmoo** [comment]
Ok, unit tests are going to be messy due to usage of fakes. One more reason for me to dislike fakes :) 

That's about as far as I can get today, and unfortunately I don't expect I will be able to get back to this for a little while. If you want to make changes on top of https://github.com/jaredmoo/elastic-db-tools/tree/NetCore please go ahead and I'll be happy to review. Just try to make each individual commit small so that it's reviewable. 

Cheers

**Comment by pristinesource** [comment]
@jaredmoo: How do you feel about me switching to xUnit for unit tests while I am updating them to work with the new changes for Net Core?

**Comment by jaredmoo** [comment]
That would be fine, but if possible would you be able to do that on a separate branch based on master (separate from net core changes)? So the branching is like
```
master -> net core       /-> net core with tests
           \->xunit          -/
```

The reason that I ask is that it would be nice to be able to merge xunit into master independently of net core. 

Also the latest xunit version that is available on my build servers is 2.1.0, will that work?

**Comment by pristinesource** [comment]
Absolutely! I will do that :)

**Comment by jaredmoo** [comment]
Awesome, thanks!

**Comment by jaredmoo** [comment]
Removed fakes dependency: https://github.com/Azure/elastic-db-tools/pull/143

**Comment by JohnGalt1717** [comment]
Are there any samples or guidance on how to setup constructors and dependency injection etc. in .NET core yet?

**Comment by grajdeanserghei** [comment]
Hi!

Are there any updates on this issue? 

@jaredmoo I can see that https://github.com/jaredmoo/elastic-db-tools contains NetCore branch and two pull requests related to it created by @pristinesource. Why it is not merged in this repository and not included in NuGet package? Do you need any help with that? 

Are there any plans to support this library? It haven't been updated since October.

**Comment by jaredmoo** [comment]
Hi @grajdeanserghei ,

The library is still 100% supported. We are responding to issue reports, answering questions on stack overflow, and major reported bugs reported will be fixed.

On why NetCore hasn't been merged: 
1. I have some open comments on the PR for xunit migration (https://github.com/Azure/elastic-db-tools/pull/144) which were not addressed. The .NET Core landscape has been changing very quickly in the last few months and it looks like MSTest now supports .NET Core, so perhaps this xunit conversion is no longer needed. I didn't merge any of @pristinesource 's code before this because he hadn't yet signed the CLA. (This is done now, so not a problem).
2. We need to make sure we get the build tooling right.
    1. I think that there were some complexities at the time related to .xproj, project.json, and .csproj, which are mostly solved now that VS2017 and we can just use .csproj for all project types.
     2. In order to make an official release we should have Authenticode signing. This is always complicated and not something that can be done by contributors like yourself.
3. The changes in jaredmoo/NetCore haven't been fully reviewed.

Now that VS2017 is out it should be easier to make progress on tooling issues. I'll play around with using VS2017 to convert these projects in the next few days and see how it goes.

Cheers

**Comment by grajdeanserghei** [comment]
Hi @jaredmoo,

Thanks for quick replay. 

@pristinesource solution contains two projects which refers same source code files. That's good approach but you still have to maintain two projects. What do you think about porting it to .NET Standard? Are there any reasons not doing that?

Regards

**Comment by jaredmoo** [comment]
Yes I think that targeting .net standard the right way to go. Ideally version 1.4 due to the compat issues in 1.5 and 1.6,

**Comment by tomkerkhove** [comment]
Is there an update on this?

**Comment by jaredmoo** [comment]
I have been waiting for the new csproj build tools and .net standard 2 in order to ease the migration. It is still work in progress but I have a branch at https://github.com/jaredmoo/elastic-db-tools/tree/NetCore where client.dll builds on .net core and all SMM tests are also built and passing. Next steps:
1. also get MSQ tests built and passing
2. fix csproj's so they build against both net451 (for perfcounter support) and netstandard

That should be all of the coding work. Then you will be able to build a .net core assembly from source. In order to produce a nuget package I will need to code sign (not to be confused with strong name signing).

Let me know if you have any feedback. Is netstandard2 ok? Using earlier netstandard is nice for platform support but would be more work from a coding and build point of view.

**Comment by tomkerkhove** [comment]
Sounds good, nice progress! I think netstandard2 should be the best fit as this comes with a lot more APIs and should be a good foundation for this library.

**Comment by natelaff** [comment]
I'm cool with .NET Standard 2.0 as the lowest targeted version for this.

**Comment by chsakell** [comment]
Is there any progress on this and if not can we have an estimation? 
I wouldn't mind to use the library without EF but till now seems impossible and there are no any working samples for .NET Core 2.0.

**Comment by cleverguy25** [comment]
Any update on this?

**Comment by cleverguy25** [comment]
I have to say it looks like this library is just not supported anymore, either from within Microsoft or the community, despite its prominent place in the Azure SQL documentation.  .Net Core is a big part of Microsoft's platform now, and now with 2.0 being released almost 3 months ago there is little excuse.

If you need help from the community, we need to know that is pending on this.  How can we help push this over the finish line?  This goes months between communication, and always appears to be "soon" or "close".

Can we get a alpha or beta package released do you can get feedback?  Seems from the last post this was at least that close.

I am planning to use this on a new project (I have used it in the past pre-.Net Core) but right now I am seriously doubting that decision.

**Comment by jaredmoo** [comment]
I'm very sorry for your frustration. I don't believe I ever stated that this work would be done "soon", and I'm sorry you had that impression.

As stated in my most recent post in this thread ( https://github.com/Azure/elastic-db-tools/issues/118#issuecomment-325182640 ), there is working code at https://github.com/jaredmoo/elastic-db-tools/tree/NetCore which you can build and use. Since then I have resolved the MSQ test failures, and just now I have pushed the code to https://github.com/Azure/elastic-db-tools/tree/NetCore

The remaining coding task is to fix csproj's so they build against both net451 (for perfcounter support) and netstandard. I don't know exactly how to do this, but I know that the Azure management SDKs are built this way so it can be figured out by looking at https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/SqlManagement/Management.Sql/Microsoft.Azure.Management.Sql.csproj 

Then there is work to get the code signed and published on nuget. This unfortunately cannot be done by community. Lack of signing is also why I have not made an alpha binary release, but luckily since it's open source you can easily grab the code and build it easily with `dotnet build`.

**Comment by erick-thompson** [comment]
I was able to get the core branch up and running, but I ran into an issue when creating the ShardMapManager. The call

`var manager = ShardMapManagerFactory.GetSqlShardMapManager("", ShardMapManagerLoadPolicy.Eager);
`
results in a Linq exception "Sequence contains no elements"

System.InvalidOperationException occurred
  HResult=0x80131509
  Message=Sequence contains no elements
  Source=System.Linq
  StackTrace:
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.GetShardMapManagerGlobalOperation.DoGlobalExecute(IStoreTransactionScope ts) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\StoreOperations\ShardMapManagerFactory\GetShardMapManagerGlobalOperation.cs:line 51
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.StoreOperationGlobal.<Do>b__10_0() in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\StoreOperations\Base\StoreOperationGlobal.cs:line 83
   at Microsoft.Azure.SqlDatabase.ElasticScale.TransientFaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ElasticScale.Common\TransientFaultHandling\Implementation\RetryPolicy.cs:line 216
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.StoreOperationGlobal.Do() in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\StoreOperations\Base\StoreOperationGlobal.cs:line 73
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.ShardMapManagerFactory.GetSqlShardMapManager(String connectionString, ShardMapManagerLoadPolicy loadPolicy, RetryBehavior retryBehavior, EventHandler`1 retryEventHandler, Boolean throwOnFailure) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\ShardMapManager\ShardMapManagerFactory.cs:line 580
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.ShardMapManagerFactory.GetSqlShardMapManager(String connectionString, ShardMapManagerLoadPolicy loadPolicy, RetryBehavior retryBehavior, EventHandler`1 retryEventHandler) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\ShardMapManager\ShardMapManagerFactory.cs:line 508
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.ShardMapManagerFactory.GetSqlShardMapManager(String connectionString, ShardMapManagerLoadPolicy loadPolicy, RetryBehavior retryBehavior) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\ShardMapManager\ShardMapManagerFactory.cs:line 463
   at Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement.ShardMapManagerFactory.GetSqlShardMapManager(String connectionString, ShardMapManagerLoadPolicy loadPolicy) in P:\external\elastic-db-tools\Src\ElasticScale.Client\ShardManagement\ShardMapManager\ShardMapManagerFactory.cs:line 442
   at SharedServices.QuantumShardMapManager.GetShardMap() in P:\quantumeventservice\SharedServices\Repositories\QuantumShardMapManager.cs:line 23
   at SharedServices.Tests.Repositories.QuantumShardMapManagerTests.Test1() in P:\quantumeventservice\SharedServices.Tests\Repositories\QuantumShardMapManagerTests.cs:line 17

It looks like it is coming out of the SqlUtils class. I see that the SQL command is embedded in ReadOnlyScripts.resx. I'm going to see if I can get it working locally, but I assume that any of the SQL inside the resources will have similar issues.

**Comment by jaredmoo** [comment]
I've been having a similar problem with the resources but I thought it was fixed. Let me take another look.

**Comment by jaredmoo** [comment]
The issue is https://github.com/Microsoft/msbuild/issues/308 . The resources build correctly in VS but not using dotnet core SDK. Investigating fixes/workarounds.

**Comment by erick-thompson** [comment]
Nice find! It there a particular reason why interned strings aren't an option? Or are you using the locales for various flavors of SQL?

**Comment by jaredmoo** [comment]
Resx was used for developer convenience, but it's not very convenient now that resx build isn't working. I just pushed an update that switches to hardcoded strings.

**Comment by erick-thompson** [comment]
Awesome, thanks. I'll pull you changes and rebuild.

**Comment by GreenAirplane** [comment]
We are starting a new big project and we would like to use elastic tools instead of our own tool. So, my question is: @jaredmoo, do you need any community help in order to speed up .net standard support?

**Comment by jaredmoo** [comment]
Just try it out and let me know if you have any issues, also it would be great to hear if everything worked :) . The only thing left to do is signing and publishing an official release, which unfortunately can't be done by community.

**Comment by GreenAirplane** [comment]
Just wanted to provide an update. We've created a "proof of concept" application and looks like it works for us! We are using Elastic Tools with EF Core and so far so good.

Btw, there was one time when some stored procedures were not created in a shard db (e.g. spBulkOperationShardMappingsLocal). However, I was not able to reproduce it again. Also, I've found issue #147 with a similar description, so I don't think that this case is related to .net standard support.

**Comment by GreenAirplane** [comment]
One more thing, it might help some other people who decide to use pre-release version.

We had a strange issue when this lib didn't work in a new solution. It just crashed with "Could not load file or assembly" error. It looks like there is an issue in .net core how works with libs from local disk. The easy way to fix is to generate a local NuGet package and add it to the new solution. Works like a miracle..

**Comment by TomBeckett** [comment]
Any update on when this may be signed / published as an official release?

**Comment by drewauman** [comment]
Can you let us know when this is going to be signed and published?   Looking forward to it!

**Comment by jaredmoo** [comment]
Thanks for the feedback and patience everyone. It is now published on nuget.
https://github.com/Azure/elastic-db-tools/releases/tag/v2.0.0-preview2
https://www.nuget.org/packages/Microsoft.Azure.SqlDatabase.ElasticScale.Client/2.0.0-preview2

**Comment by RJM76** [comment]
Are any Entity Framework Core samples available?

I'm using 2.0.0-preview2 from Nuget but getting stuck on Database.SetInitializer<MyDbContext<T>>(...) in both the SetInitializerForConnection and CreateDDRConnection methods.  This is because the samples aren't using EF Core so System.Data.Entity is not available.  And in EF Core it appears the equivalent is DatabaseFacade, although they do not line up in terms if initialization.

For example, calling 'Database.SetInitializer<MyDbContext<T>>(null);' from CreateDDRConnection results in a "An object reference is required for the non-static field, method, or property 'DbContext.Database'" error.

**Comment by jaredmoo** [comment]
Sorry I haven't gotten a chance to try with EF core :(

**Comment by bbqchickenrobot** [comment]
In case anyone was wondering I whipped a quick EFCore sample to use with preview-2.0.0 build.... it's not perfect and/or tested - but if you wanted to understand the concept here it is: 

```
using System;
using System.Configuration;
using System.Data.SqlClient;
using System.Linq;
using Microsoft.Azure.SqlDatabase.ElasticScale.ShardManagement;
using Microsoft.EntityFrameworkCore;

namespace eftest
{
    // todo - convert public accessors to appropriate access level
    public static class MultiTenantConfigurator
    {
        public static MyContext GetContext(int shardKey) =>
            new MyContext(
                GetDbContextOptions<MyContext>(shardKey, "shardMapManager", "My_ShardMap", "MyDbConnStringName"));

        public static DbContextOptions<TContext> GetDbContextOptions<TContext>(int shardKey,
            string shardMapConnstringName, string shardMapName, string dbConnString, int commandTimeout = 600)
            where TContext : DbContext => 
            GetDbContextOptions<int, TContext>(shardKey, shardMapConnstringName, shardMapName, dbConnString, commandTimeout);

        public static DbContextOptions<TContext> GetDbContextOptions<Tkey, TContext>(Tkey shardKey, 
            string shardMapConnstringName, string shardMapName, string dbConnString, int commandTimeout = 600) 
            where TContext : DbContext
        {
            var builder = new DbContextOptionsBuilder<TContext>();
            var conn = OpenDDRConnection(GetShardMapManager(GetConnectionString(shardMapConnstringName), shardMapName),
                shardKey, GetConnectionString(dbConnString));
            var options = builder.UseSqlServer(conn, b => b.CommandTimeout(commandTimeout));
            return options.Options;
        }

        public static string GetConnectionString(string name) => ConfigurationManager.ConnectionStrings[name].ConnectionString;

        public static ShardMap GetShardMapManager(string connString, string shardMapName)
        {
            var smm = ShardMapManagerFactory.GetSqlShardMapManager(connString, ShardMapManagerLoadPolicy.Lazy);
            var sm = smm.GetListShardMap<int>(shardMapName);
            return sm;
        }

        /// <summary>
        /// Wrapper function for ShardMap.OpenConnectionForKey() that automatically sets CONTEXT_INFO to the correct
        /// tenantId before returning a connection. As a best practice, you should only open connections using this 
        /// method to ensure that CONTEXT_INFO is always set before executing a query.
        /// </summary>
        public static SqlConnection OpenDDRConnection<T>(ShardMap shardMap, T shardingKey, string connString)
        {
            // No initialization
            //Database.SetInitializer<MyContext>(null);

            // Ask shard map to broker a validated connection for the given key
            SqlConnection conn = null;
            try
            {
                connString = MassageConnString(connString);
                conn = shardMap.OpenConnectionForKey(shardingKey, connString, ConnectionOptions.Validate);

                // Set TenantId in SESSION_CONTEXT to shardingKey to enable Row-Level Security filtering
                var cmd = conn.CreateCommand();

                cmd.CommandText = @"exec sp_set_session_context @key=N'TenantId', @value=@shardingKey";
                cmd.Parameters.AddWithValue("@shardingKey", shardingKey);
                cmd.ExecuteNonQuery();

                return conn;
            }
            catch (Exception)
            {
                conn?.Dispose();
                throw;
            }


            string MassageConnString(string connstring)  // local function
            {
                var list = connstring.Split(';').ToList();
                list.Remove(list.First(x => x.StartsWith("Data Source")));
                list.Remove(list.First(x => x.StartsWith("Initial Catalog")));
                return string.Join(";", list);
            }
        }
    }
}
```

Hope that helps....

**Comment by drewauman** [comment]
Could anyone send a full working sample of sharding with EF 7 Core?   I have to admit, I have been banging on this thing off and on for way too long.   Would be very helpful if someone has a working sample.   Would be a huge help here ...    I want multiple Azure database in an elastic pool, multiple tenants per database with the ability to filter at the database level and to do split/merge to move tenants from database to database as needed.  Help!  :)  Thank you ahead of time, I appreciate the help.  Feel free to send it directly to me at drewauman [at] outlook.com if you could.  Thanks!!

**Comment by bbqchickenrobot** [comment]
@drewauman  -  I provided a full working sample.  Create your DbContext from your database, create your shards and then you can use my gist to talk with it.  I don't have a working sample as this would require to create a database, configure sharding (this is done within azure sql), etc....

**Comment by drewauman** [comment]
Thank BBQ .... I have got this working several times in the past using Entity Framework.    It's Entity Framework "Core" that I have had a hard time finding documentation and/or examples for.   I have my solution setup, I just seem to be having issues creating the schema the first time in the shardlet DBs .... it is driving me crazy!  ha ha and it's probably something simple that is causing it.   That's why I was asking for a working version (I was just looking for the code or solution).       

The downloadable example that is provided in the link that you sent is entity framework 6 it appears, not core.   Thoughts?    Thanks for your time sir!

**Comment by drewauman** [comment]
I can't seem to figure out how to add migrations for the first time or kick off migrations for the first time and then subsequent times into the shard database.   I am in a Entity Framework Core Code First setup.   I need to initialize migrations for the first time and then migrate automatically every time new migrations are added.   Thoughts?

**Comment by SychevIgor** [comment]
Thank you for everything you did!
It will be nice to have a version without "preview" in the end
https://github.com/Azure/elastic-db-tools/tags here we can find only preview-2 that was shipped in February 2018... 8th month ago.

**Comment by natelaff** [comment]
Why was this closed when there is no final version and no documentation w/ EF Core? Kind of shaking my confidence a little bit in starting a new project with this.

**Comment by vigouredelaruse** [comment]
with the imminent move to a write once run almost anywhere .net 5 situation, what is the current assessment of the relevance of porting this library to .net core as per https://docs.microsoft.com/en-us/dotnet/core/porting/ ?

i fear time spent with .net framework dependencies merely leaves this open source effort subject to the predations of a branched commercial effort that supports .net core

please advise
